stacks-signer/src/coordinator.rs

```diff
@@ -0,0 +1,230 @@
+use slog::slog_debug;
+use stacks_common::debug;
+use stacks_common::types::chainstate::ConsensusHash;
+use stacks_common::util::hash::Sha256Sum;
+use wsts::curve::ecdsa;
+use wsts::state_machine::PublicKeys;
+
+/// TODO: test this value and adjust as necessary. Maybe make configurable?
+pub const COORDINATOR_OPERATION_TIMEOUT_SECS: u64 = 300;
+
+/// TODO: test this value and adjust as necessary. Maybe make configurable?
+pub const COORDINATOR_TENURE_TIMEOUT_SECS: u64 = 600;
+
+/// The coordinator selector
+#[derive(Clone, Debug)]
+pub struct CoordinatorSelector {
+    /// The ordered list of potential coordinators for a specific consensus hash
+    coordinator_ids: Vec<u32>,
+    /// The current coordinator id
+    coordinator_id: u32,
+    /// The current coordinator index into the coordinator ids list
+    coordinator_index: usize,
+    /// The last message received time for the current coordinator
+    pub last_message_time: Option<Instant>,
+    /// The time the coordinator started its tenure
+    tenure_start: Instant,
+    /// The public keys of the coordinators
+    public_keys: PublicKeys,
+}
+
+impl From<PublicKeys> for CoordinatorSelector {
+    /// Create a new Coordinator selector from the given list of public keys
+    fn from(public_keys: PublicKeys) -> Self {
+        let coordinator_ids =
+            CoordinatorSelector::calculate_coordinator_ids(&public_keys, &ConsensusHash::empty());
+        let coordinator_id = *coordinator_ids
+            .first()
+            .expect("FATAL: No registered signers");
+        let coordinator_index = 0;
+        let last_message_time = None;
+        let tenure_start = Instant::now();
+        Self {
+            coordinator_ids,
+            coordinator_id,
+            coordinator_index,
+            last_message_time,
+            tenure_start,
+            public_keys,
+        }
+    }
+}
+
+impl CoordinatorSelector {
+    /// Update the coordinator id
+    fn update_coordinator(&mut self, new_coordinator_ids: Vec<u32>) {
+        self.last_message_time = None;
+        self.coordinator_index = if new_coordinator_ids != self.coordinator_ids {
+            // We have advanced our block height and should select from the new list
+            let mut new_index: usize = 0;
+            self.coordinator_ids = new_coordinator_ids;
+            let new_coordinator_id = *self
+                .coordinator_ids
+                .first()
+                .expect("FATAL: No registered signers");
+            if new_coordinator_id == self.coordinator_id {
+                // If the newly selected coordinator is the same as the current and we have more than one available, advance immediately to the next
+                if self.coordinator_ids.len() > 1 {
+                    new_index = new_index.saturating_add(1);
+                }
+            }
+            new_index
+        } else {
+            let mut new_index = self.coordinator_index.saturating_add(1);
```

Comment:
Could simplify this with `self.coordinator_index = self.coordinator_index.saturating_add(1) % self.coordinator_ids.len()`, assuming that the list of coordinator IDs is always positive.

Reply:
ah yes true. It is always greater than 0. It will panic earlier with a FATAL: no registered signers